features/filesystem/bd/BufferedBlockDevice.cpp

```diff
@@ -0,0 +1,229 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "BufferedBlockDevice.h"
+#include "mbed_assert.h"
+#include <algorithm>
+#include <string.h>
+
+static inline uint32_t align_down(bd_size_t val, bd_size_t size)
+{
+    return val / size * size;
+}
+
+BufferedBlockDevice::BufferedBlockDevice(BlockDevice *bd)
+    : _bd(bd), _bd_program_size(0), _curr_aligned_addr(0), _flushed(true), _cache(0)
+{
+}
+
+BufferedBlockDevice::~BufferedBlockDevice()
+{
+    deinit();
+    if (_cache) {
+        delete[] _cache;
+    }
+}
+
+int BufferedBlockDevice::init()
+{
+    int err = _bd->init();
+    if (err) {
+        return err;
+    }
+
+    _bd_program_size = _bd->get_program_size();
+
+    if (!_cache) {
+        _cache = new uint8_t[_bd_program_size];
+    }
+
+    _curr_aligned_addr = _bd->size();
+    _flushed = true;
+
+    return 0;
+}
+
+int BufferedBlockDevice::deinit()
+{
+    return _bd->deinit();
```

Comment:
Should this free the cache for symmetry? Seems reasonable if they've explicitly called `deinit` and allocation's in `init`, but maybe there are practical reasons not to.

Reply:
Well, in this specific block device, there's no difference between freeing the cache (or any other resource for that matter) here or in the destructor. However, tried to conform to other block devices, which let the data live through init/deinit calls.

Reply:
Ah! The heap block device is special in this case. (It's "pretending" the heap is non-volatile storage.) I would suggest freeing in deinit also.

Reply:
Gotcha. Moved to deinit.